scarlet/observation.py

```diff
@@ -176,41 +180,50 @@ def match(self, model_frame):
-            psf_shape = np.array(self.frame.psfs[0].shape) + self._padding
-            shape = np.array(model_frame.shape[1:]) + psf_shape - 1
+            psf_shape = np.array(self.frame.psfs.shape)
+            psf_shape[1:] += self._padding
+            shape = np.array(model_frame.shape) + psf_shape - 1
+            shape[0] = model_frame.shape[0]
+
             # Choose the optimal shape for FFTPack DFT
-            self._fftpack_shape = [fftpack.helper.next_fast_len(d) for d in shape]
+            self._fftpack_shape = [fftpack.helper.next_fast_len(d) for d in shape[1:]]
             # autograd.numpy.fft does not currently work
             # if the last dimension is odd
             while self._fftpack_shape[-1] % 2 != 0:
                 _shape = self._fftpack_shape[-1] + 1
                 self._fftpack_shape[-1] = fftpack.helper.next_fast_len(_shape)
+
             # Store the pre-fftpack optimization slices
-            self._slices = tuple([slice(s) for s in shape])
+            self.slices = tuple(([slice(s) for s in shape]))
 
             # Now we setup the parameters for the psf -> kernel FFTs
-            shape = np.array(model_frame.psfs[0].shape) + np.array(self.frame.psfs[0].shape) - 1
-            _fftpack_shape = [fftpack.helper.next_fast_len(d) for d in shape]
-            # Deconvolve the target PSF
-            target_fft = np.fft.rfftn(model_frame.psfs[0], _fftpack_shape)
-
-            # Match the PSF in each band
-            _diff_kernels_fft = []
-            for psf in self.frame.psfs:
-                _psf_fft = np.fft.rfftn(psf, _fftpack_shape)
-                kernel = np.fft.fftshift(np.fft.irfftn(_psf_fft / target_fft, _fftpack_shape))
-                kernel *= model_frame.psfs[0].sum()
-                if kernel.shape[0] % 2 == 0:
-                    kernel = kernel[1:, 1:]
-                kernel = _centered(kernel, psf_shape)
-                _diff_kernels_fft.append(np.fft.rfftn(kernel, self._fftpack_shape))
+            shape = np.array(model_frame.psfs.shape) - 1
+            shape[1:] += np.array(self.frame.psfs[0].shape)
+
+            _fftpack_shape = [fftpack.helper.next_fast_len(d) for d in shape[1:]]
+
+            # fft of the target psf
+            target_fft = np.fft.rfftn(model_frame.psfs, _fftpack_shape, axes=(1, 2))
+
+            # fft of the observation's PSFs in each band
+            _psf_fft = np.fft.rfftn(self.psfs, _fftpack_shape, axes=(1, 2))
+
+            # Diff kernel between observation and target psf in Fourrier
+            kernels = np.fft.ifftshift(np.fft.irfftn(_psf_fft / target_fft, _fftpack_shape, axes=(1, 2)), axes=(1, 2))
+
+            if kernels.shape[1] % 2 == 0:
+                kernels = kernels[:, 1:, 1:]
+
+            kernels = _centered(kernels, psf_shape)
+
+            _diff_kernels_fft = np.fft.rfftn(kernels, self._fftpack_shape, axes=(1, 2))
 
             self._diff_kernels_fft = np.array(_diff_kernels_fft)
```

Comment:
why is this copy necessary? you should be able to use the result of the `ffftn` in line 226.

Reply:
Indeed